In [1]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv("DataCoSupplyChainDataset.csv", encoding='latin1')
#why encoding ? => DataCo dataset me customer names, city names, ya product names me kuch special characters ho sakte hain

In [9]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [10]:
df.shape

(180519, 53)

In [11]:
df.columns

Index(['Type', 'Days for shipping (real)', 'Days for shipment (scheduled)',
       'Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Id', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Email', 'Customer Fname', 'Customer Id',
       'Customer Lname', 'Customer Password', 'Customer Segment',
       'Customer State', 'Customer Street', 'Customer Zipcode',
       'Department Id', 'Department Name', 'Latitude', 'Longitude', 'Market',
       'Order City', 'Order Country', 'Order Customer Id',
       'order date (DateOrders)', 'Order Id', 'Order Item Cardprod Id',
       'Order Item Discount', 'Order Item Discount Rate', 'Order Item Id',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Order Zipcode', 'Product Card Id', 'Product Category Id',
       'Product De

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [13]:
df.describe()

,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Late_delivery_risk,Category Id,Customer Id,Customer Zipcode,Department Id,Latitude,...,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Price,Product Status
count,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180516.000000,180519.000000,180519.000000,...,180519.000000,180519.000000,180519.000000,180519.000000,24840.000000,180519.000000,180519.000000,0.0,180519.000000,180519.0
mean,3.497654,2.931847,21.974989,183.107609,0.548291,31.851451,6691.379495,35921.126914,5.443460,29.719955,...,2.127638,203.772096,183.107609,21.974989,55426.132327,692.509764,31.851451,NaN,141.232550,0.0
std,1.623722,1.374449,104.433526,120.043670,0.497664,15.640064,4162.918106,37542.461122,1.629246,9.813646,...,1.453451,132.273077,120.043670,104.433526,31919.279101,336.446807,15.640064,NaN,139.732492,0.0
min,0.000000,0.000000,-4274.979980,7.490000,0.000000,2.000000,1.000000,603.000000,2.000000,-33.937553,...,1.000000,9.990000,7.490000,-4274.979980,1040.000000,19.000000,2.000000,NaN,9.990000,0.0
25%,2.000000,2.000000,7.000000,104.379997,0.000000,18.000000,3258.500000,725.000000,4.000000,18.265432,...,1.000000,119.980003,104.379997,7.000000,23464.000000,403.000000,18.000000,NaN,50.000000,0.0
50%,3.000000,4.000000,31.520000,163.990005,1.000000,29.000000,6457.000000,19380.000000,5.000000,33.144863,...,1.000000,199.919998,163.990005,31.520000,59405.000000,627.000000,29.000000,NaN,59.990002,0.0
75%,5.000000,4.000000,64.800003,247.399994,1.000000,45.000000,9779.000000,78207.000000,7.000000,39.279617,...,3.000000,299.950012,247.399994,64.800003,90008.000000,1004.000000,45.000000,NaN,199.990005,0.0
max,6.000000,4.000000,911.799988,1939.989990,1.000000,76.000000,20757.000000,99205.000000,12.000000,48.781933,...,5.000000,1999.989990,1939.989990,911.799988,99301.000000,1363.000000,76.000000,NaN,1999.989990,0.0


In [14]:
df.describe(include='object')

,Type,Delivery Status,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Lname,Customer Password,Customer Segment,...,Order City,Order Country,order date (DateOrders),Order Region,Order State,Order Status,Product Image,Product Name,shipping date (DateOrders),Shipping Mode
count,180519,180519,180519,180519,180519,180519,180519,180511,180519,180519,...,180519,180519,180519,180519,180519,180519,180519,180519,180519,180519
unique,4,4,50,563,2,1,782,1109,1,3,...,3597,164,65752,23,1089,9,118,118,63701,4
top,DEBIT,Late delivery,Cleats,Caguas,EE. UU.,XXXXXXXXX,Mary,Smith,XXXXXXXXX,Consumer,...,Santo Domingo,Estados Unidos,12/14/2016 12:29,Central America,Inglaterra,COMPLETE,http://images.acmesports.sports/Perfect+Fitnes...,Perfect Fitness Perfect Rip Deck,1/5/2016 5:58,Standard Class
freq,69295,98977,24551,66770,111146,180519,65150,64104,180519,93504,...,2211,24840,5,28341,6722,59491,24515,24515,10,107752


In [15]:
df.isnull().sum()

Type                                  0
Days for shipping (real)              0
Days for shipment (scheduled)         0
Benefit per order                     0
Sales per customer                    0
Delivery Status                       0
Late_delivery_risk                    0
Category Id                           0
Category Name                         0
Customer City                         0
Customer Country                      0
Customer Email                        0
Customer Fname                        0
Customer Id                           0
Customer Lname                        8
Customer Password                     0
Customer Segment                      0
Customer State                        0
Customer Street                       0
Customer Zipcode                      3
Department Id                         0
Department Name                       0
Latitude                              0
Longitude                             0
Market                                0


In [16]:
df.duplicated().sum()

np.int64(0)

In [17]:
df.nunique()

Type                                  4
Days for shipping (real)              7
Days for shipment (scheduled)         4
Benefit per order                 21998
Sales per customer                 2927
Delivery Status                       4
Late_delivery_risk                    2
Category Id                          51
Category Name                        50
Customer City                       563
Customer Country                      2
Customer Email                        1
Customer Fname                      782
Customer Id                       20652
Customer Lname                     1109
Customer Password                     1
Customer Segment                      3
Customer State                       46
Customer Street                    7458
Customer Zipcode                    995
Department Id                        11
Department Name                      11
Latitude                          11250
Longitude                          4487
Market                                5


**EXPLORATORY SUMMARY**

row = 180519 
columns = 53 
Duplicate Rows = 0  
missing columns = Customer Zipcode, Customer Lname, Order Zipcode, Product Description 
Numeric Columns = 29
categorical columns = 24
Date Columns = order date (DateOrders), shipping date (DateOrders)


In [25]:
df.isnull().sum().sort_values(ascending=False).head(6)

Product Description       180519
Order Zipcode             155679
Customer Lname                 8
Customer Zipcode               3
Type                           0
Order Profit Per Order         0
dtype: int64

In [26]:
# we will drop prouct description and order zipcode as they have more missing values

In [27]:
df = df.drop('Product Description', axis=1)

In [30]:
df = df.drop('Order Zipcode',axis=1)

In [31]:
df.shape

(180519, 51)

In [33]:
#droping Lname as it is not usefull for us

In [34]:
df = df.drop('Customer Lname',axis=1)

In [35]:
#we will also drop zipcode as we already have city state and country

In [37]:
df = df.drop('Customer Zipcode',axis=1)

In [38]:
df.shape

(180519, 49)

In [39]:
#make date time columns

In [40]:
df['order date (DateOrders)'] = pd.to_datetime(df['order date (DateOrders)'])
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'])

In [45]:
df['order date (DateOrders)'].head(3)

0   2018-01-31 22:56:00
1   2018-01-13 12:27:00
2   2018-01-13 12:06:00
Name: order date (DateOrders), dtype: datetime64[ns]

In [46]:
df['shipping date (DateOrders)'].head(3)

0   2018-02-03 22:56:00
1   2018-01-18 12:27:00
2   2018-01-17 12:06:00
Name: shipping date (DateOrders), dtype: datetime64[ns]

**Before Cleaning, One More Exploration Step**

In [47]:
df.nunique().sort_values()

Product Status                        1
Customer Password                     1
Customer Email                        1
Customer Country                      2
Late_delivery_risk                    2
Customer Segment                      3
Type                                  4
Shipping Mode                         4
Delivery Status                       4
Days for shipment (scheduled)         4
Order Item Quantity                   5
Market                                5
Days for shipping (real)              7
Order Status                          9
Department Id                        11
Department Name                      11
Order Item Discount Rate             18
Order Region                         23
Customer State                       46
Category Name                        50
Category Id                          51
Product Category Id                  51
Product Price                        75
Order Item Product Price             75
Product Image                       118


In [54]:
#dropping unneccessary columns
df = df.drop(columns=[
    'Customer Password',
    'Customer Email',
    'Customer Fname',
    'Customer Street',
    'Product Image'
])


In [57]:
df =df.drop('Product Status',axis=1)

In [58]:
df.nunique().sort_values()

Customer Country                      2
Late_delivery_risk                    2
Customer Segment                      3
Type                                  4
Delivery Status                       4
Shipping Mode                         4
Days for shipment (scheduled)         4
Order Item Quantity                   5
Market                                5
Days for shipping (real)              7
Order Status                          9
Department Id                        11
Department Name                      11
Order Item Discount Rate             18
Order Region                         23
Customer State                       46
Category Name                        50
Category Id                          51
Product Category Id                  51
Product Price                        75
Order Item Product Price             75
Product Card Id                     118
Order Item Cardprod Id              118
Product Name                        118
Order Item Profit Ratio             162


In [59]:
df.shape

(180519, 43)

In [61]:
#category distribution
for col in df.select_dtypes(include='object'):
    print(f"\n--- {col} ---")
    print(df[col].value_counts().head())


--- Type ---
Type
DEBIT       69295
TRANSFER    49883
PAYMENT     41725
CASH        19616
Name: count, dtype: int64

--- Delivery Status ---
Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

--- Category Name ---
Category Name
Cleats                  24551
Men's Footwear          22246
Women's Apparel         21035
Indoor/Outdoor Games    19298
Fishing                 17325
Name: count, dtype: int64

--- Customer City ---
Customer City
Caguas         66770
Chicago         3885
Los Angeles     3417
Brooklyn        3412
New York        1816
Name: count, dtype: int64

--- Customer Country ---
Customer Country
EE. UU.        111146
Puerto Rico     69373
Name: count, dtype: int64

--- Customer Segment ---
Customer Segment
Consumer       93504
Corporate      54789
Home Office    32226
Name: count, dtype: int64

--- Customer State ---
Customer State
PR    69373
CA    29223
NY    11327
TX     

In [62]:
df.describe().T

,count,mean,min,25%,50%,75%,max,std
Days for shipping (real),180519.0,3.497654,0.0,2.0,3.0,5.0,6.0,1.623722
Days for shipment (scheduled),180519.0,2.931847,0.0,2.0,4.0,4.0,4.0,1.374449
Benefit per order,180519.0,21.974989,-4274.97998,7.0,31.52,64.800003,911.799988,104.433526
Sales per customer,180519.0,183.107609,7.49,104.379997,163.990005,247.399994,1939.98999,120.04367
Late_delivery_risk,180519.0,0.548291,0.0,0.0,1.0,1.0,1.0,0.497664
Category Id,180519.0,31.851451,2.0,18.0,29.0,45.0,76.0,15.640064
Customer Id,180519.0,6691.379495,1.0,3258.5,6457.0,9779.0,20757.0,4162.918106
Department Id,180519.0,5.44346,2.0,4.0,5.0,7.0,12.0,1.629246
Latitude,180519.0,29.719955,-33.937553,18.265432,33.144863,39.279617,48.781933,9.813646
Longitude,180519.0,-84.915675,-158.025986,-98.446312,-76.847908,-66.370583,115.263077,21.433241


In [63]:
#memory usage
df.memory_usage(deep=True).sum() / 1024**2

np.float64(204.23495960235596)

**🚀Data Cleaning**

In [65]:
categorical_cols = [
    'Delivery Status',
    'Shipping Mode',
    'Customer Segment',
    'Market',
    'Order Status'
]

for col in categorical_cols:
    print(f"\n-----{col}-----")
    print(df[col].value_counts())


-----Delivery Status-----
Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

-----Shipping Mode-----
Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

-----Customer Segment-----
Customer Segment
Consumer       93504
Corporate      54789
Home Office    32226
Name: count, dtype: int64

-----Market-----
Market
LATAM           51594
Europe          50252
Pacific Asia    41260
USCA            25799
Africa          11614
Name: count, dtype: int64

-----Order Status-----
Order Status
COMPLETE           59491
PENDING_PAYMENT    39832
PROCESSING         21902
PENDING            20227
CLOSED             19616
ON_HOLD             9804
SUSPECTED_FRAUD     4062
CANCELED            3692
PAYMENT_REVIEW      1893
Name: count, dtype: int64


**🚀Saving this data**

In [67]:
df.to_csv('cleaned_supply_chain.csv', index=False)